### ISSUE TITLE
[Bug]: Problematic warnings during export of Spkrec-Ecapa-voxceleb to ONNX (opset=10). 

### ISSUE THREAD

**Issue by EnlNovius** [speechbrain/speechbrain#1750]
### Describe the bug

For a real time project, I need to convert [Spkrec-Ecapa-voxceleb](https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb) to ONNX in order to use it in cpp with [OnnxRuntime](https://onnxruntime.ai/).

Since FFT is supported by `torch.onnx.export` (https://github.com/pytorch/pytorch/issues/44822), I exported the network in 3 blocks (FBank, TDNN and AttentiveStatisticsPooling).

When exporting the last block (AttentiveStatisticsPooling) with pytorch, I got two warnings (see the **To Reproduce** part for the code):

```bash
conda/speechbrain/lib/python3.9/site-packages/speechbrain/dataio/dataio.py:704: TracerWarning: Using len to get tensor shape might cause the trace to be incorrect. Recommended usage would be tensor.shape[0]. Passing a tensor of different shape might lead to errors or silently give incorrect results.
  ).expand(len(length), max_len) < length.unsqueeze(1)
conda/speechbrain/lib/python3.9/site-packages/speechbrain/dataio/dataio.py:712: TracerWarning: torch.as_tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  mask = torch.as_tensor(mask, dtype=dtype, device=device)
```

The generated model passes the tests of the onnx python library:
```python
import onnx
model = onnx.load("attentive_statistics_pooling.onnx")
onnx.checker.check_model(model)
```

However, if I try to open this model in cpp with onnxruntime:

```cpp
_session = Ort::Session(env, model_filepath.c_str(), session_options);
```

I have the following error that appears at runtime:

```shell
what():  Load model from attentive_statistics_pooling.onnx failed:This is an invalid model. Type Error: Type 'tensor(float)' of input parameter (/asp/Unsqueeze_4_output_0) of operator (Equal) in node (/asp/Equal_1) is invalid
```

I couldn't find a correct solution on the cpp side to get the network working, so I looked at the two warnings. The first one is apparently benign, but after exploring, solving the second one fixes the problem (Once corrected, the Model is correctly loaded in cpp). The solution I found is to modify the `length_to_mask` function in the `dataio/dataio.py` file:

```python
def length_to_mask(length, max_len=None, dtype=None, device=None):
    """Creates a binary mask for each sequence.

    Reference: https://discuss.pytorch.org/t/how-to-generate-variable-length-mask/23397/3

    Arguments
    ---------
    length : torch.LongTensor
        Containing the length of each sequence in the batch. Must be 1D.
    max_len : int
        Max length for the mask, also the size of the second dimension.
    dtype : torch.dtype, default: None
        The dtype of the generated mask.
    device: torch.device, default: None
        The device to put the mask variable.

    Returns
    -------
    mask : tensor
        The binary mask.

    Example
    -------
    >>> length=torch.Tensor([1,2,3])
    >>> mask=length_to_mask(length)
    >>> mask
    tensor([[1., 0., 0.],
            [1., 1., 0.],
            [1., 1., 1.]])
    """
    assert len(length.shape) == 1

    # Current code:
    # if max_len is None:
    #     max_len = length.max().long().item()  # using arange to generate mask
    # mask = torch.arange(
    #     max_len, device=length.device, dtype=length.dtype
    # ).expand(len(length), max_len) < length.unsqueeze(1)
    # if dtype is None:
    #     dtype = length.dtype
    #
    # if device is None:
    #     device = length.device
    #
    # mask = torch.as_tensor(mask, dtype=dtype, device=device)
    
    # Proposition
    if dtype is None:
        dtype = length.dtype

    if device is None:
        device = length.device

    if max_len is None:
        max_len = length.max().long().item()
        
    mask = torch.arange(
        max_len, device=device, dtype=dtype # Use directly the device and dtype here.
    ).expand(length.shape[0], max_len) < length.unsqueeze(1) # Since we are modifying the function, we can also replace the len with a shape[0] to avoid the 1st warning.

    return mask
```

Is this the right solution or do you have another solution to make the export work properly?

### Expected behaviour

Export to ONNX with `torch.onnx.export` without the warnings:

```bash
conda/speechbrain/lib/python3.9/site-packages/speechbrain/dataio/dataio.py:712: TracerWarning: torch.as_tensor results are registered as constants in the trace. 
You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. 
In any other case, this might cause the trace to be incorrect.
  mask = torch.as_tensor(mask, dtype=dtype, device=device)
```

Which seems to cause problems when the model is then imported into C++

### To Reproduce

```bash
conda install -c anaconda protobuf notebook
pip install speechbrain
pip install onnx
```

```python
import torch
from speechbrain.pretrained import EncoderClassifier

device = None

if not device:
    device = "cuda:0" if torch.cuda.is_available() else "cpu"

run_opts = {"device": device}

embed_model = EncoderClassifier.from_hparams(source="speechbrain/spkrec-ecapa-voxceleb",
                                             savedir="pretrained_models/spkrec-ecapa-voxceleb",
                                             run_opts=run_opts)

class AttentiveStatisticsPooling(torch.nn.Module):
    def __init__(self, asp, asp_bn, fc):
        super().__init__()
        self.asp = asp
        self.asp_bn = asp_bn
        self.fc = fc

    def forward(self, wav_embeddings):
        wav_embeddings = torch.transpose(wav_embeddings, 1, 2)
        embedding = self.asp(wav_embeddings)
        embedding = self.asp_bn(embedding)
        embedding = self.fc(embedding)
        return embedding

dummy_input = torch.randn((1, 98, 3072), requires_grad=True, device=device) # 98 is a random value

sub_model = AttentiveStatisticsPooling(embed_model.mods.embedding_model.asp, 
                                       embed_model.mods.embedding_model.asp_bn, 
                                       embed_model.mods.embedding_model.fc)

sub_model.eval()

torch.onnx.export(sub_model,
                  dummy_input,
                  "attentive_statistics_pooling.onnx",
                  export_params=True,
                  opset_version=10,
                  do_constant_folding=True,
                  input_names = ['input'],
                  output_names = ['output'],
                  dynamic_axes={'input' : {0: 'batch_size', 1 : 'windows'},
                                'output': {0: 'batch_size'}})
```

Output:
```bash
conda/speechbrain/lib/python3.9/site-packages/speechbrain/dataio/dataio.py:704: TracerWarning: Using len to get tensor shape might cause the trace to be incorrect. Recommended usage would be tensor.shape[0]. Passing a tensor of different shape might lead to errors or silently give incorrect results.
  ).expand(len(length), max_len) < length.unsqueeze(1)
conda/speechbrain/lib/python3.9/site-packages/speechbrain/dataio/dataio.py:712: TracerWarning: torch.as_tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  mask = torch.as_tensor(mask, dtype=dtype, device=device)
```

### Versions

python 3.9
speechbrain               0.5.13
torch                     1.13.0

**Comment by neso613** [comment]
Hi @EnlNovius 
I was able to convert pretrained [classifier model](https://github.com/speechbrain/speechbrain/tree/develop/recipes/VoxLingua107/lang_id#inference)(language identification) into onnx successfully.

But I am not able to understand its preprocessing part. 
If you help me to guide how to remove torch dependency while its onnx inferencing,its very appreciated.


I am sharing here the code used - 

import onnxruntime
ort_session = onnxruntime.InferenceSession("lid_speechbrain_pretrained.onnx")

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

signal = classifier.load_audio(path)   # This classifier model is pertrained PyTorch and used here to create embedding only.
emb =  classifier.encode_batch(signal)

ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(emb)}
ort_outs= ort_session.run(None, ort_inputs)
ort_outs = torch.from_numpy(ort_outs[0][0])
score_, index_ = torch.max(ort_outs, dim=-1)

f = open('pretrained_models/lang-id-commonlanguage_ecapa/label_encoder.ckpt','r')
languages = []
text_lab_ = ''
languages = f.readlines()
for i in languages:
    if str(int(index_)) in i:
        text_lab_ = i.split('=>')[0]
        break
print("ONNX:    ",score_, index_,text_lab_)

**Comment by EnlNovius** [comment]
Hello @neso613,

The original pytorch model is composed of 3 large blocks. 
- A FBank (with a default window of 25ms = 400 values and a step of 10ms = 160 values)
- A TDNN applied on the FBank results, generating n intermediate embeddings (3072 values per embedding)
- An Attentive Statistics Pooling block to merge the n TDNN results into a single embedding of size 192.

The block that cannot be exported as is in ONNX is the FBank (~MFCC) which uses a STFT. The FBank is composed of 3 steps: application of a hamming window, the STFT and application of the filter. To do it without PyTorch, you have to apply the hamming window and the STFT with a third party library on the input signal, then apply the filtering.

Then you have to apply the TDNN (TDNNBlock) to compute n intermediate embeddings from the Fbank result.

Finally we can apply on these n blocks the attention block (the one which causes problems).

Summary of the structure:

```python
├─ModuleDict: 1                                    []                        --
|    └─Fbank: 2-1                                  [-1, 10, 80]              --
|    |    └─STFT: 3-1                              [-1, 10, 201, 2]          --
|    |    └─Filterbank: 3-2                        [-1, 10, 80]              --
|    └─InputNormalization: 2-2                     [-1, 10, 80]              --
|    └─ECAPA_TDNN: 2-3                             [-1, 1, 192]              --
|    |    └─TDNNBlock: 3-3                         [-1, 3072, 11]            (9,446,400)
|    |    └─AttentiveStatisticsPooling: 3-4        [-1, 6144, 1]             (1,576,320)
|    |    └─BatchNorm1d: 3-5                       [-1, 6144, 1]             (12,288)
|    |    └─Conv1d: 3-6                            [-1, 192, 1]              (1,179,840)
|    └─Classifier: 2-4                             [-1, 1, 7205]             (1,383,360)
```

I hope this answers the question =)

**Comment by neso613** [comment]
@EnlNovius Yes, I got the answer. Thanks!

**Comment by mrg79433283** [comment]
I am not very familiar with speechtrain. Follow your prompts, I was able to convert Spkrec-Ecapa-voxceleb into onnx successfully. But I  wonder how to integrate the model into the speaker recognition module. 
thank you.

**Comment by marziye-A** [comment]
> Hi @EnlNovius I was able to convert pretrained [classifier model](https://github.com/speechbrain/speechbrain/tree/develop/recipes/VoxLingua107/lang_id#inference)(language identification) into onnx successfully.
> 
> But I am not able to understand its preprocessing part. If you help me to guide how to remove torch dependency while its onnx inferencing,its very appreciated.
> 
> I am sharing here the code used -
> 
> import onnxruntime ort_session = onnxruntime.InferenceSession("lid_speechbrain_pretrained.onnx")
> 
> def to_numpy(tensor): return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
> 
> signal = classifier.load_audio(path) # This classifier model is pertrained PyTorch and used here to create embedding only. emb = classifier.encode_batch(signal)
> 
> ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(emb)} ort_outs= ort_session.run(None, ort_inputs) ort_outs = torch.from_numpy(ort_outs[0][0]) score_, index_ = torch.max(ort_outs, dim=-1)

dear @neso613 , I also want to onnx the language identification model but I faced errors, can you please share your code?
